goalie/function_data.py

```diff
@@ -30,7 +29,14 @@ def __init__(self, time_partition, function_spaces):
         self.time_partition = time_partition
         self.function_spaces = function_spaces
         self._data = None
-        self.labels = self._label_dict["steady"] + self._label_dict["unsteady"]
+        self.labels = (
+            [
+                [label for label in self._label_dict[field]]
+                for field in self.time_partition.field_types
+            ][0]
+            if self.time_partition.fields
+            else []
+        )
```

Comment:
I'm really confused by this. Is it not sufficient to just do the following?
```
self.labels = self._label_dict["steady" if time_partition.steady else "unsteady"]
```

Reply:
@jwallwork23 - I had missed the `steady` attribute. Yes, that does significantly simplify things. See commit #https://github.com/pyroteus/goalie/pull/149/commits/32119bfc7c58609d6bf95f3e27465112084f94fb

Reply:
Oh actually it's not quite so simple because (as @ddundo has pointed out previously) some fields might be "steady" and some might be "unsteady". So I don't think we can just set `self.labels` for all cases in this way. Instead, it should be determined per-field.

Apologies! So some refactoring will be required. Does what I said here sound right to you both?

Reply:
Ah yes, I should have also pointed that out in #116... sorry! That looks like it complicates `_data_by_label` in this mixed case, where we have both steady-unsteady fields, since unsteady fields will have labels that steady ones don't. Although, in this mixed case the "steady" fields aren't actually steady, so the notion of "forward_old" and "adjoint_next" still makes sense for them, right?

The reason why we introduced field types was to not get the no dependency error in `MeshSeq._dependency`. So I don't actually see a reason why we would care about field type anywhere else... maybe I shouldn't have pointed this out in the first place.

Reply:
@ddundo - I think it would help to have a simple mixed case test. Do you have one already implemented elsewhere in the code base I can use as a go-by?

Reply:
It looks like in `time_partition.py` the `field_types` variable is set for all fields as `unsteady` as long as there is more than one subinterval or more than one timestep. If it is just one interval with one timestep, all fields are set to `steady`. 

As @ddundo pointed out, in a mixed case, there would be more than one timestep by default so even the `steady` field in a mixed case would need to be handled over all the timesteps. At the moment, the labels `forward_old` and/or `adjoint_next` would still apply to the `steady` component of the mixed case to handle it over more than one timestep.

@jwallwork23 I am not sure that further refactoring is needed, given the current setup? I would like to first implement a mixed `steady`, `unsteady` case to see where the current setup fails.

Reply:
Thanks @acse-ej321, that would indeed be useful! I could probably simplify my ice evolution problem, but maybe someone at the meeting today will be able to propose a better example :)